core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,198 @@
+    ValidationException.check(ref.isTag(), "Ref with name %s is a tag not a branch", name);
+    SnapshotRef updatedRef = SnapshotRef.builderFrom(ref, snapshotId).build();
+    updatedRefs.put(name, updatedRef);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation setMinSnapshotsToKeep(String name, int minSnapshotsToKeep) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such ref with name %s", name);
+    ValidationException.check(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    SnapshotRef updateBranch = SnapshotRef.builderFrom(ref)
+        .minSnapshotsToKeep(minSnapshotsToKeep)
+        .build();
+    updatedRefs.put(name, updateBranch);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation setMaxSnapshotAgeMs(String name, long maxSnapshotAgeMs) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such ref with name %s", name);
+    ValidationException.check(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    SnapshotRef updateBranch = SnapshotRef.builderFrom(ref)
+        .maxSnapshotAgeMs(maxSnapshotAgeMs)
+        .build();
+    updatedRefs.put(name, updateBranch);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation setMaxRefAgeMs(String name, long maxRefAgeMs) {
+    Preconditions.checkNotNull(name, "Reference name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such ref with name %s", name);
+    SnapshotRef updatedRef = SnapshotRef.builderFrom(ref)
+        .maxRefAgeMs(maxRefAgeMs)
+        .build();
+    updatedRefs.put(name, updatedRef);
+    return this;
+  }
+
+  private TableMetadata internalApply() {
+    TableMetadata.Builder updatedBuilder = TableMetadata.buildFrom(base);
+    Map<String, SnapshotRef> refDiff = refDiff(base.refs(), updatedRefs);
+    for (Map.Entry<String, SnapshotRef> diffEntry : refDiff.entrySet()) {
+      String refName = diffEntry.getKey();
+      if (diffEntry.getValue() == null) {
+        SnapshotRef ref = base.ref(refName);
+        removeRef(updatedBuilder, refName, ref);
+      } else {
+        SnapshotRef ref = updatedRefs.get(diffEntry.getKey());
+        if (ref.isBranch()) {
+          updatedBuilder.setBranch(refName, ref);
+        } else {
+          updatedBuilder.setTag(refName, ref);
+        }
+      }
+    }
+    return updatedBuilder.build();
+  }
+
+  private void removeRef(TableMetadata.Builder updatedBuilder, String refName, SnapshotRef ref) {
+    if (ref.isBranch()) {
+      updatedBuilder.removeBranch(refName);
+    } else {
+      updatedBuilder.removeTag(refName);
+    }
+  }
+
+  private Map<String, SnapshotRef> refDiff(Map<String, SnapshotRef> currRefs, Map<String, SnapshotRef> newRefs) {
```

Comment:
Since I think `apply` should return the full updated map of references, I think this could be merged into `internalApply`. That would probably be more straightforward than returning a map and then interpreting it.

Reply:
Good point, this will simplify it